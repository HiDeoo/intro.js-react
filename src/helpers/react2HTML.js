import React from 'react';
import { renderToStaticMarkup } from 'react-dom/server'

function react2HTML(intro: React.ElementType) {
  return renderToStaticMarkup(intro);  
}

export default react2HTML;
