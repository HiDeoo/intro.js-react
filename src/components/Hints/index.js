import introJs from 'intro.js';
import PropTypes from 'prop-types';
import { Component } from 'react';

import * as introJsPropTypes from '../../helpers/proptypes';
import * as introJsDefaultProps from '../../helpers/defaultProps';
import { isServer } from '../../helpers/server';

/**
 * Intro.js Hints Component.
 */
export default class Hints extends Component {
  /**
   * React Props
   * @type {Object}
   */
  static propTypes = {
    enabled: PropTypes.bool,
    hints: PropTypes.arrayOf(
      PropTypes.shape({
        element: PropTypes.string.isRequired,
        hint: PropTypes.string.isRequired,
        hintPosition: introJsPropTypes.hintPosition,
      })
    ).isRequired,
    onClick: PropTypes.func,
    onClose: PropTypes.func,
    options: introJsPropTypes.options,
  };

  /**
   * React Default Props
   * @type {Object}
   */
  static defaultProps = {
    enabled: false,
    onClick: null,
    onClose: null,
    options: introJsDefaultProps.options,
  };

  /**
   * Creates a new instance of the component.
   * @class
   * @param {Object} props - The props of the component.
   */
  constructor(props) {
    super(props);

    this.introJs = null;
    this.isConfigured = false;

    this.installIntroJs();
  }

  /**
   * Lifecycle: componentDidMount.
   * We use this event to enable Intro.js hints at mount time if enabled right from the start.
   */
  componentDidMount() {
    if (this.props.enabled) {
      this.configureIntroJs();
      this.renderHints();
    }
  }

  /**
   * Lifecycle: componentDidUpdate.
   * @param  {Object} prevProps - The previous props.
   */
  componentDidUpdate(prevProps) {
    const { enabled, hints, options } = this.props;

    if (!this.isConfigured || prevProps.hints !== hints || prevProps.options !== options) {
      this.configureIntroJs();
      this.renderHints();
    }

    if (prevProps.enabled !== enabled) {
      this.renderHints();
    }
  }

  /**
   * Lifecycle: componentWillUnmount.
   * We use this even to hide the hints when the component is unmounted.
   */
  componentWillUnmount() {
    this.introJs.hideHints();
  }

  /**
   * Installs Intro.js.
   */
  installIntroJs() {
    if (isServer()) {
      return;
    }

    this.introJs = introJs();

    const { onClick, onClose } = this.props;

    if (onClick) {
      this.introJs.onhintclick(onClick);
    }

    if (onClose) {
      this.introJs.onhintclose(onClose);
    }
  }

  /**
   * Configures Intro.js if not already configured.
   */
  configureIntroJs() {
    const { options, hints } = this.props;

    // We need to remove all hints otherwise new hints won't be added.
    this.introJs.removeHints();

    this.introJs.setOptions({ ...options, hints });

    this.isConfigured = true;
  }

  /**
   * Renders the Intro.js hints.
   */
  renderHints() {
    const { enabled, hints } = this.props;

    if (enabled && hints.length > 0) {
      this.introJs.showHints();
    } else if (!enabled) {
      this.introJs.hideHints();
    }
  }

  /**
   * Renders the component.
   * @return {null} We do not want to render anything.
   */
  render() {
    return null;
  }
}
