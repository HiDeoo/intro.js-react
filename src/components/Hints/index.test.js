import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import Hints from './index';
import * as server from '../../helpers/server';

/**
 * Hints.
 * @type {Hint[]}
 */
const hints = [
  {
    element: '.test',
    hint: 'test',
  },
];

describe('Hints', () => {
  test('should render nothing', () => {
    const tree = renderer.create(<Hints enabled hints={[]} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('should not configure Intro.js at mount time if not enabled', () => {
    const wrapper = shallow(<Hints hints={hints} />, {
      lifecycleExperimental: true,
    });

    expect(wrapper.instance().isConfigured).toBe(false);
  });

  test('should configure Intro.js at mount time if enabled at least once', () => {
    const wrapper = shallow(<Hints enabled hints={hints} />, {
      lifecycleExperimental: true,
    });

    wrapper.setProps({ enabled: false });

    expect(wrapper.instance().isConfigured).toBe(true);
  });

  test('should configure Intro.js when enabled if not already configured', () => {
    const wrapper = shallow(<Hints hints={hints} />, {
      lifecycleExperimental: true,
    });

    wrapper.instance().configureIntroJs = jest.fn();
    wrapper.update();
    wrapper.setProps({ enabled: true });

    expect(wrapper.instance().configureIntroJs).toHaveBeenCalled();
  });

  test('should call renderHints when enabled while being disabled', () => {
    const wrapper = shallow(<Hints hints={hints} />, {
      lifecycleExperimental: true,
    });

    wrapper.instance().renderHints = jest.fn();
    wrapper.update();
    wrapper.setProps({ enabled: true });

    expect(wrapper.instance().renderHints).toHaveBeenCalled();
  });

  test('should call renderHints when enabled while being already enabled', () => {
    const wrapper = shallow(<Hints enabled hints={hints} />, {
      lifecycleExperimental: true,
    });

    wrapper.instance().renderHints = jest.fn();
    wrapper.update();
    wrapper.setProps({ enabled: false });

    expect(wrapper.instance().renderHints).toHaveBeenCalled();
  });

  test('should call renderHints everytime the enabled prop is modified', () => {
    const wrapper = shallow(<Hints hints={hints} />, {
      lifecycleExperimental: true,
    });

    wrapper.instance().renderHints = jest.fn();
    wrapper.update();
    wrapper.setProps({ enabled: true });
    wrapper.setProps({ enabled: false });
    wrapper.setProps({ enabled: true });

    expect(wrapper.instance().renderHints).toHaveBeenCalledTimes(4);

    wrapper.unmount();
  });

  test('should not call renderHints when the enabled, hints or options are not modified', () => {
    const wrapper = shallow(<Hints enabled hints={hints} />, {
      lifecycleExperimental: true,
    });

    wrapper.instance().renderHints = jest.fn();
    wrapper.update();
    wrapper.setProps({ enabled: true, hints });

    expect(wrapper.instance().renderHints).not.toHaveBeenCalled();
  });

  test('should register the onClick callback', () => {
    const onClick = jest.fn();

    const wrapper = shallow(<Hints hints={hints} onClick={onClick} />);

    expect(wrapper.instance().introJs.onHintClick).toBe(onClick);
  });

  test('should register the onClose callback', () => {
    const onClose = jest.fn();

    const wrapper = shallow(<Hints hints={hints} onClose={onClose} />);

    expect(wrapper.instance().introJs.onHintClose).toBe(onClose);
  });

  test('should not install intro.js during SSR', () => {
    jest.spyOn(server, 'isServer').mockReturnValueOnce(true);

    const wrapper = shallow(<Hints hints={hints} />);

    expect(wrapper.instance().introJs).toBe(null);
  });
});
