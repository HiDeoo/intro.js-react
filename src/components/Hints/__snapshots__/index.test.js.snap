// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hints should render jsx 1`] = `null`;

exports[`Hints should render nothing 1`] = `null`;
